### PR TITLE
Fix parameterized runner used with parameters of Android classes

### LINKED ISSUE #871: Parameterized Runner
I did some tests to get one of my tests that uses `ParameterizedRobolectricTestRunner` to work without crashing. I originally was getting argument type mismatch when it tried to use the test's constructor which took an `enum` I defined (`Weekday`) and a `boolean`. I believe I have troubleshooted it down to being an issue with `enum`. This is a test that worked before with a regular `ParameterizedTestRunner`, but I later added some Android dependencies to the class, so I needed to switch to Robolectric runner. Is it possible that there's some issue with how the `enum` parameter is being used in the Robolectric test runner?

If you care, my solution was to use the int value of my enum in the `@Parameters` and convert it back to the enum in the test. And admittedly, this is an `enum` that will likely get converted to constants with static helper functions, but that's later on my To Do list.

### PR DESCRIPTION
**Pull request by WonderCsabo**
Fixes #871.

### CODE REVIEW

**Comment by WonderCsabo** [comment]
@erd @jongerrish any chance this gets reviewed sooner or later? Unfortunetaly we cannot update to Robolectrc 3 due to the bug, so i would really like to see this included in 3.1.

File: robolectric/src/main/java/org/robolectric/ParameterizedRobolectricTestRunner.java [hunk 1]
```diff
@@ -98,6 +102,16 @@ protected void validateConstructor(List<Throwable> errors) {
     }
 
     @Override
+    protected Statement methodBlock(FrameworkMethod method, Config config, AndroidManifest appManifest, SdkEnvironment sdkEnvironment) {
+      configureShadows(sdkEnvironment, config);
+
+      DeepCloner deepCloner = new DeepCloner(sdkEnvironment.getRobolectricClassLoader());
+      parameters = deepCloner.clone(parameters);
```
Comment (erd) [inline robolectric/src/main/java/org/robolectric/ParameterizedRobolectricTestRunner.java]: Why does the classloader need to be deep-cloned? Is there any way to do this without introducing `xstream` as a dependency?

Comment (WonderCsabo) [inline robolectric/src/main/java/org/robolectric/ParameterizedRobolectricTestRunner.java]: I do not think there is another way.

The problem here is that the parameter objects are created with the system class loader, but the test is later run by the Robolectric class loader. You cannot event pass the parameters to the test class constructor - you will get an error.

Another solution would be calling the parameter generator method again, but with the Robolectric class loader. I think there are two problems with that:
* we should not call that method more than once during a test, this is not what the user would expect
* if the generator is random, or time-dependent, different parameters will show up in the test names and the test

Also please note that not the class-loader is deep-cloned, the parameter objects are.

**Comment by erd** [comment]
Could you rebase this? I'll merge afterwards.

**Comment by WonderCsabo** [comment]
Rebased.

We can add some optimization to this. Currently, [here](https://github.com/WonderCsabo/robolectric/commit/b8676de8a3ed4766c20e4eac615209a7f337643b#diff-67bb0637a1d0ba8c39d2260e55d129c5R109) all parameters are cloned, regardless it is actually needed or not. We could iterate over the array, and only clone the parameter if `InstrumentationConfiguration.shouldAcquire()` returns `true`. What do you think?

**Comment by erd** [comment]
> What do you think?

If you benchmark it and it provides a nice, tangible benefit, I think it would be worthwhile. Otherwise, I think the current implementation is fine.

**Comment by WonderCsabo** [comment]
I tested it, with just looking at the test run time, i could not measure significant benefit. Do you guys use any good Java profiler?

**Pull request merged**